File: test-errors/src/transformers/CallExpressionTransformer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallExpressionTransformer = void 0;
const path_1 = __importDefault(require("path"));
const AssertTransformer_1 = require("./features/AssertTransformer");
const CreateAssertTransformer_1 = require("./features/CreateAssertTransformer");
const CreateIsTransformer_1 = require("./features/CreateIsTransformer");
const CreateRandomTransformer_1 = require("./features/CreateRandomTransformer");
const CreateValidateTransformer_1 = require("./features/CreateValidateTransformer");
const IsTransformer_1 = require("./features/IsTransformer");
const RandomTransformer_1 = require("./features/RandomTransformer");
const ValidateTransformer_1 = require("./features/ValidateTransformer");
const JsonApplicationTransformer_1 = require("./features/json/JsonApplicationTransformer");
const JsonAssertParseTransformer_1 = require("./features/json/JsonAssertParseTransformer");
const JsonAssertStringifyTransformer_1 = require("./features/json/JsonAssertStringifyTransformer");
const JsonCreateAssertParseTransformer_1 = require("./features/json/JsonCreateAssertParseTransformer");
const JsonCreateAssertStringifyTransformer_1 = require("./features/json/JsonCreateAssertStringifyTransformer");
const JsonCreateIsParseTransformer_1 = require("./features/json/JsonCreateIsParseTransformer");
const JsonCreateIsStringifyTransformer_1 = require("./features/json/JsonCreateIsStringifyTransformer");
const JsonCreateStringifyTransformer_1 = require("./features/json/JsonCreateStringifyTransformer");
const JsonCreateValidateParseTransformer_1 = require("./features/json/JsonCreateValidateParseTransformer");
const JsonCreateValidateStringifyProgrammer_1 = require("./features/json/JsonCreateValidateStringifyProgrammer");
const JsonIsParseTransformer_1 = require("./features/json/JsonIsParseTransformer");
const JsonIsStringifyTransformer_1 = require("./features/json/JsonIsStringifyTransformer");
const JsonStringifyTransformer_1 = require("./features/json/JsonStringifyTransformer");
const JsonValidateParseTransformer_1 = require("./features/json/JsonValidateParseTransformer");
const JsonValidateStringifyTransformer_1 = require("./features/json/JsonValidateStringifyTransformer");
const MetadataTransformer_1 = require("./features/misc/MetadataTransformer");
const MiscAssertCloneTransformer_1 = require("./features/misc/MiscAssertCloneTransformer");
const MiscAssertPruneTransformer_1 = require("./features/misc/MiscAssertPruneTransformer");
const MiscCloneTransformer_1 = require("./features/misc/MiscCloneTransformer");
const MiscCreateAssertCloneTransformer_1 = require("./features/misc/MiscCreateAssertCloneTransformer");
const MiscCreateAssertPruneTransformer_1 = require("./features/misc/MiscCreateAssertPruneTransformer");
const MiscCreateCloneTransformer_1 = require("./features/misc/MiscCreateCloneTransformer");
const MiscCreateIsCloneTransformer_1 = require("./features/misc/MiscCreateIsCloneTransformer");
const MiscCreateIsPruneTransformer_1 = require("./features/misc/MiscCreateIsPruneTransformer");
const MiscCreatePruneTransformer_1 = require("./features/misc/MiscCreatePruneTransformer");
const MiscCreateValidateCloneTransformer_1 = require("./features/misc/MiscCreateValidateCloneTransformer");
const MiscCreateValidatePruneTransformer_1 = require("./features/misc/MiscCreateValidatePruneTransformer");
const MiscIsCloneTransformer_1 = require("./features/misc/MiscIsCloneTransformer");
const MiscIsPruneTransformer_1 = require("./features/misc/MiscIsPruneTransformer");
const MiscLiteralsTransformer_1 = require("./features/misc/MiscLiteralsTransformer");
const MiscPruneTransformer_1 = require("./features/misc/MiscPruneTransformer");
const MiscValidateCloneTransformer_1 = require("./features/misc/MiscValidateCloneTransformer");
const MiscValidatePruneTransformer_1 = require("./features/misc/MiscValidatePruneTransformer");
const CreateProtobufAssertDecodeTransformer_1 = require("./features/protobuf/CreateProtobufAssertDecodeTransformer");
const CreateProtobufAssertEncodeTransformer_1 = require("./features/protobuf/CreateProtobufAssertEncodeTransformer");
const CreateProtobufDecodeTransformer_1 = require("./features/protobuf/CreateProtobufDecodeTransformer");
const CreateProtobufEncodeTransformer_1 = require("./features/protobuf/CreateProtobufEncodeTransformer");
const CreateProtobufIsDecodeTransformer_1 = require("./features/protobuf/CreateProtobufIsDecodeTransformer");
const CreateProtobufIsEncodeTransformer_1 = require("./features/protobuf/CreateProtobufIsEncodeTransformer");
const CreateProtobufValidateDecodeTransformer_1 = require("./features/protobuf/CreateProtobufValidateDecodeTransformer");
const CreateProtobufValidateEncodeTransformer_1 = require("./features/protobuf/CreateProtobufValidateEncodeTransformer");
const ProtobufAssertDecodeTransformer_1 = require("./features/protobuf/ProtobufAssertDecodeTransformer");
const ProtobufAssertEncodeTransformer_1 = require("./features/protobuf/ProtobufAssertEncodeTransformer");
const ProtobufDecodeTransformer_1 = require("./features/protobuf/ProtobufDecodeTransformer");
const ProtobufEncodeTransformer_1 = require("./features/protobuf/ProtobufEncodeTransformer");
const ProtobufIsDecodeTransformer_1 = require("./features/protobuf/ProtobufIsDecodeTransformer");
const ProtobufIsEncodeTransformer_1 = require("./features/protobuf/ProtobufIsEncodeTransformer");
const ProtobufMessageTransformer_1 = require("./features/protobuf/ProtobufMessageTransformer");
const ProtobufValidateDecodeTransformer_1 = require("./features/protobuf/ProtobufValidateDecodeTransformer");
const ProtobufValidateEncodeTransformer_1 = require("./features/protobuf/ProtobufValidateEncodeTransformer");
var CallExpressionTransformer;
(function (CallExpressionTransformer) {
    CallExpressionTransformer.transform = (project) => (expression) => {
        //----
        // VALIDATIONS
        //----
        // SIGNATURE DECLARATION
        const declaration = project.checker.getResolvedSignature(expression)?.declaration;
        if (!declaration)
            return expression;
        // FILE PATH
        const location = path_1.default.resolve(declaration.getSourceFile().fileName);
        if (isTarget(location) === false)
            return expression;
        //----
        // TRANSFORMATION
        //----
        // FUNCTION NAME
        const module = location
            .split(path_1.default.sep)
            .at(-1)
            .split(".")[0];
        const { name } = project.checker.getTypeAtLocation(declaration).symbol;
        // FIND TRANSFORMER
        const functor = FUNCTORS[module]?.[name];
        if (functor === undefined)
            return expression;
        // RETURNS WITH TRANSFORMATION
        const result = functor()(project)(expression.expression)(expression);
        return result ?? expression;
    };
    const isTarget = (location) => {
        const files = Object.keys(FUNCTORS);
        return files.some((f) => location.includes(path_1.default.join("node_modules", "typia", "lib", `${f}.d.ts`)) ||
            location ===
                path_1.default.resolve(path_1.default.join(__dirname, "..", `${f}.ts`)) ||
            location ===
                path_1.default.resolve(path_1.default.join(__dirname, "..", "..", "..", "src", `${f}.ts`)));
    };
})(CallExpressionTransformer || (exports.CallExpressionTransformer = CallExpressionTransformer = {}));
const FUNCTORS = {
    module: {
        // BASIC
        assert: () => AssertTransformer_1.AssertTransformer.transform(false),
        assertType: () => AssertTransformer_1.AssertTransformer.transform(false),
        is: () => IsTransformer_1.IsTransformer.transform(false),
        validate: () => ValidateTransformer_1.ValidateTransformer.transform(false),
        // STRICT
        assertEquals: () => AssertTransformer_1.AssertTransformer.transform(true),
        equals: () => IsTransformer_1.IsTransformer.transform(true),
        validateEquals: () => ValidateTransformer_1.ValidateTransformer.transform(true),
        // RANDOM + INTERNAL
        random: () => RandomTransformer_1.RandomTransformer.transform,
        metadata: () => (P) => () => MetadataTransformer_1.MetadataTransformer.transform(P),
        // FACTORIES
        createAssert: () => CreateAssertTransformer_1.CreateAssertTransformer.transform(false),
        createAssertType: () => CreateAssertTransformer_1.CreateAssertTransformer.transform(false),
        createIs: () => CreateIsTransformer_1.CreateIsTransformer.transform(false),
        createValidate: () => CreateValidateTransformer_1.CreateValidateTransformer.transform(false),
        createAssertEquals: () => CreateAssertTransformer_1.CreateAssertTransformer.transform(true),
        createEquals: () => CreateIsTransformer_1.CreateIsTransformer.transform(true),
        createValidateEquals: () => CreateValidateTransformer_1.CreateValidateTransformer.transform(true),
        createRandom: () => CreateRandomTransformer_1.CreateRandomTransformer.transform,
    },
    json: {
        // SCHEMA
        application: () => (P) => () => JsonApplicationTransformer_1.JsonApplicationTransformer.transform(P),
        // PARSER
        isParse: () => JsonIsParseTransformer_1.JsonIsParseTransformer.transform,
        assertParse: () => JsonAssertParseTransformer_1.JsonAssertParseTransformer.transform,
        validateParse: () => JsonValidateParseTransformer_1.JsonValidateParseTransformer.transform,
        // STRINGIFY
        stringify: () => JsonStringifyTransformer_1.JsonStringifyTransformer.transform,
        assertStringify: () => JsonAssertStringifyTransformer_1.JsonAssertStringifyTransformer.transform,
        isStringify: () => JsonIsStringifyTransformer_1.JsonIsStringifyTransformer.transform,
        validateStringify: () => JsonValidateStringifyTransformer_1.JsonValidateStringifyTransformer.transform,
        // FACTORIES
        createIsParse: () => JsonCreateIsParseTransformer_1.JsonCreateIsParseTransformer.transform,
        createAssertParse: () => JsonCreateAssertParseTransformer_1.JsonCreateAssertParseTransformer.transform,
        createValidateParse: () => JsonCreateValidateParseTransformer_1.JsonCreateValidateParseTransformer.transform,
        createStringify: () => JsonCreateStringifyTransformer_1.JsonCreateStringifyTransformer.transform,
        createAssertStringify: () => JsonCreateAssertStringifyTransformer_1.JsonCreateAssertStringifyTransformer.transform,
        createIsStringify: () => JsonCreateIsStringifyTransformer_1.JsonCreateIsStringifyTransformer.transform,
        createValidateStringify: () => JsonCreateValidateStringifyProgrammer_1.JsonCreateValidateStringifyTransformer.transform,
    },
    protobuf: {
        // SCHEMA
        message: () => ProtobufMessageTransformer_1.ProtobufMessageTransformer.transform,
        // ENCODE
        encode: () => ProtobufEncodeTransformer_1.ProtobufEncodeTransformer.transform,
        assertEncode: () => ProtobufAssertEncodeTransformer_1.ProtobufAssertEncodeTransformer.transform,
        isEncode: () => ProtobufIsEncodeTransformer_1.ProtobufIsEncodeTransformer.transform,
        validateEncode: () => ProtobufValidateEncodeTransformer_1.ProtobufValidateEncodeTransformer.transform,
        // DECODE
        decode: () => ProtobufDecodeTransformer_1.ProtobufDecodeTransformer.transform,
        assertDecode: () => ProtobufAssertDecodeTransformer_1.ProtobufAssertDecodeTransformer.transform,
        isDecode: () => ProtobufIsDecodeTransformer_1.ProtobufIsDecodeTransformer.transform,
        validateDecode: () => ProtobufValidateDecodeTransformer_1.ProtobufValidateDecodeTransformer.transform,
        // FACTORIES
        createEncode: () => CreateProtobufEncodeTransformer_1.CreateProtobufEncodeTransformer.transform,
        createAssertEncode: () => CreateProtobufAssertEncodeTransformer_1.CreateProtobufAssertEncodeTransformer.transform,
        createIsEncode: () => CreateProtobufIsEncodeTransformer_1.CreateProtobufIsEncodeTransformer.transform,
        createValidateEncode: () => CreateProtobufValidateEncodeTransformer_1.CreateProtobufValidateEncodeTransformer.transform,
        createDecode: () => CreateProtobufDecodeTransformer_1.CreateProtobufDecodeTransformer.transform,
        createAssertDecode: () => CreateProtobufAssertDecodeTransformer_1.CreateProtobufAssertDecodeTransformer.transform,
        createIsDecode: () => CreateProtobufIsDecodeTransformer_1.CreateProtobufIsDecodeTransformer.transform,
        createValidateDecode: () => CreateProtobufValidateDecodeTransformer_1.CreateProtobufValidateDecodeTransformer.transform,
    },
    misc: {
        literals: () => (project) => () => MiscLiteralsTransformer_1.MiscLiteralsTransformer.transform(project),
        // CLONE
        clone: () => MiscCloneTransformer_1.MiscCloneTransformer.transform,
        assertClone: () => MiscAssertCloneTransformer_1.MiscAssertCloneTransformer.transform,
        isClone: () => MiscIsCloneTransformer_1.MiscIsCloneTransformer.transform,
        validateClone: () => MiscValidateCloneTransformer_1.MiscValidateCloneTransformer.transform,
        // PRUNE
        prune: () => MiscPruneTransformer_1.MiscPruneTransformer.transform,
        assertPrune: () => MiscAssertPruneTransformer_1.MiscAssertPruneTransformer.transform,
        isPrune: () => MiscIsPruneTransformer_1.MiscIsPruneTransformer.transform,
        validatePrune: () => MiscValidatePruneTransformer_1.MiscValidatePruneTransformer.transform,
        // FACTORIES
        createClone: () => MiscCreateCloneTransformer_1.MiscCreateCloneTransformer.transform,
        createAssertClone: () => MiscCreateAssertCloneTransformer_1.MiscCreateAssertCloneTransformer.transform,
        createIsClone: () => MiscCreateIsCloneTransformer_1.MiscCreateIsCloneTransformer.transform,
        createValidateClone: () => MiscCreateValidateCloneTransformer_1.MiscCreateValidateCloneTransformer.transform,
        createPrune: () => MiscCreatePruneTransformer_1.MiscCreatePruneTransformer.transform,
        createAssertPrune: () => MiscCreateAssertPruneTransformer_1.MiscCreateAssertPruneTransformer.transform,
        createIsPrune: () => MiscCreateIsPruneTransformer_1.MiscCreateIsPruneTransformer.transform,
        createValidatePrune: () => MiscCreateValidatePruneTransformer_1.MiscCreateValidatePruneTransformer.transform,
    },
};

File: packages/transform/src/CallExpressionTransformer.go
package transform

import (
	"path/filepath"
	"strings"

	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/samchon/ttsc/packages/ttsc/driver"
	features "github.com/samchon/typia/packages/transform/src/features"
	httpfeatures "github.com/samchon/typia/packages/transform/src/features/http"
	jsonfeatures "github.com/samchon/typia/packages/transform/src/features/json"
	llmfeatures "github.com/samchon/typia/packages/transform/src/features/llm"
	miscfeatures "github.com/samchon/typia/packages/transform/src/features/misc"
	notationfeatures "github.com/samchon/typia/packages/transform/src/features/notations"
	protobuffeatures "github.com/samchon/typia/packages/transform/src/features/protobuf"
	reflectfeatures "github.com/samchon/typia/packages/transform/src/features/reflect"
	internal "github.com/samchon/typia/packages/transform/src/internal"
)

type ITransformCallExpressionProps struct {
	File *ast.SourceFile
	Text string
}

var CallExpressionTransformer = callExpressionTransformerNamespace{}

type callExpressionTransformerNamespace struct{}

func (callExpressionTransformerNamespace) transform(props ITransformCallExpressionProps) ITransformNodeResult {
	roots := ImportTransformer.typiaRoots(props.Text)
	if len(roots) == 0 {
		return ITransformNodeResult{}
	}
	calls := scanTypiaCalls(props.Text, roots)
	out := ITransformNodeResult{Rewrites: make([]driver.Rewrite, 0, len(calls))}
	for _, call := range calls {
		if invalidTypiaCall(call) {
			continue
		}
		task := functor(call)
		if task == nil {
			continue
		}
		replacement := task(call)
		if replacement == "" {
			continue
		}
		out.Rewrites = append(out.Rewrites, driver.Rewrite{
			File:          props.File,
			RootName:      call.Root,
			Namespaces:    call.Namespaces,
			Method:        call.Method,
			Replacement:   replacement,
			ConsumeParens: call.Factory,
		})
	}
	return out
}

func invalidTypiaCall(call internal.ITypiaCallExpression) bool {
	typ := call.TypeText
	route := filepath.ToSlash(strings.Join(append(append([]string{}, call.Namespaces...), call.Method), "/"))
	if strings.Contains(typ, "WeakMap<") || strings.Contains(typ, "WeakSet<") {
		return true
	}
	if strings.Contains(typ, "bigint") && strings.HasPrefix(route, "json/") {
		return true
	}
	if strings.Contains(typ, "any") && strings.HasPrefix(route, "protobuf/") {
		return true
	}
	if strings.Contains(typ, "&") && !strings.Contains(typ, "tags.") && !strings.Contains(typ, "typia.tags.") {
		return true
	}
	return false
}

func functor(call internal.ITypiaCallExpression) internal.Task {
	module := "module"
	if len(call.Namespaces) != 0 {
		module = call.Namespaces[0]
	}
	if table, ok := FUNCTORS[module]; ok {
		return table[call.Method]
	}
	return nil
}

var FUNCTORS = map[string]map[string]internal.Task{
	"module": {
		"assert":                  features.AssertTransformer.Transform(false, false),
		"assertGuard":             features.AssertTransformer.Transform(false, true),
		"assertType":              features.AssertTransformer.Transform(false, false),
		"is":                      features.IsTransformer.Transform(false),
		"validate":                features.ValidateTransformer.Transform(false),
		"assertEquals":            features.AssertTransformer.Transform(true, false),
		"assertGuardEquals":       features.AssertTransformer.Transform(true, true),
		"equals":                  features.IsTransformer.Transform(true),
		"validateEquals":          features.ValidateTransformer.Transform(true),
		"random":                  features.RandomTransformer.Transform,
		"metadata":                reflectfeatures.ReflectMetadataTransformer.Transform,
		"createAssert":            features.CreateAssertTransformer.Transform(false, false),
		"createAssertGuard":       features.CreateAssertTransformer.Transform(false, true),
		"createAssertType":        features.CreateAssertTransformer.Transform(false, false),
		"createIs":                features.CreateIsTransformer.Transform(false),
		"createValidate":          features.CreateValidateTransformer.Transform(false),
		"createAssertEquals":      features.CreateAssertTransformer.Transform(true, false),
		"createAssertGuardEquals": features.CreateAssertTransformer.Transform(true, true),
		"createEquals":            features.CreateIsTransformer.Transform(true),
		"createValidateEquals":    features.CreateValidateTransformer.Transform(true),
		"createRandom":            features.CreateRandomTransformer.Transform,
	},
	"http": {
		"parameter": httpfeatures.HttpParameterTransformer.Transform,
		"query":     httpfeatures.HttpQueryTransformer.Transform,
	},
	"json": {
		"schema":                  jsonfeatures.JsonSchemaTransformer.Transform,
		"schemas":                 jsonfeatures.JsonSchemasTransformer.Transform,
		"stringify":               jsonfeatures.JsonStringifyTransformer.Transform,
		"assertStringify":         jsonfeatures.JsonAssertStringifyTransformer.Transform,
		"isStringify":             jsonfeatures.JsonIsStringifyTransformer.Transform,
		"validateStringify":       jsonfeatures.JsonValidateStringifyTransformer.Transform,
		"assertParse":             jsonfeatures.JsonAssertParseTransformer.Transform,
		"isParse":                 jsonfeatures.JsonIsParseTransformer.Transform,
		"validateParse":           jsonfeatures.JsonValidateParseTransformer.Transform,
		"createStringify":         jsonfeatures.JsonCreateStringifyTransformer.Transform,
		"createAssertStringify":   jsonfeatures.JsonCreateAssertStringifyTransformer.Transform,
		"createIsStringify":       jsonfeatures.JsonCreateIsStringifyTransformer.Transform,
		"createValidateStringify": jsonfeatures.JsonCreateValidateStringifyTransformer.Transform,
		"createAssertParse":       jsonfeatures.JsonCreateAssertParseTransformer.Transform,
		"createIsParse":           jsonfeatures.JsonCreateIsParseTransformer.Transform,
		"createValidateParse":     jsonfeatures.JsonCreateValidateParseTransformer.Transform,
	},
	"llm": {
		"controller":       llmfeatures.LlmControllerTransformer.Transform,
		"application":      llmfeatures.LlmApplicationTransformer.Transform,
		"structuredOutput": llmfeatures.LlmStructuredOutputTransformer.Transform,
		"parameters":       llmfeatures.LlmParametersTransformer.Transform,
		"schema":           llmfeatures.LlmSchemaTransformer.Transform,
		"parse":            llmfeatures.LlmParseTransformer.Transform,
		"coerce":           llmfeatures.LlmCoerceTransformer.Transform,
		"createParse":      llmfeatures.LlmCreateParseTransformer.Transform,
		"createCoerce":     llmfeatures.LlmCreateCoerceTransformer.Transform,
	},
	"misc": {
		"literals":            miscfeatures.MiscLiteralsTransformer.Transform,
		"prune":               miscfeatures.MiscPruneTransformer.Transform,
		"assertPrune":         miscfeatures.MiscAssertPruneTransformer.Transform,
		"isPrune":             miscfeatures.MiscIsPruneTransformer.Transform,
		"validatePrune":       miscfeatures.MiscValidatePruneTransformer.Transform,
		"clone":               miscfeatures.MiscCloneTransformer.Transform,
		"assertClone":         miscfeatures.MiscAssertCloneTransformer.Transform,
		"isClone":             miscfeatures.MiscIsCloneTransformer.Transform,
		"validateClone":       miscfeatures.MiscValidateCloneTransformer.Transform,
		"createPrune":         miscfeatures.MiscCreatePruneTransformer.Transform,
		"createAssertPrune":   miscfeatures.MiscCreateAssertPruneTransformer.Transform,
		"createIsPrune":       miscfeatures.MiscCreateIsPruneTransformer.Transform,
		"createValidatePrune": miscfeatures.MiscCreateValidatePruneTransformer.Transform,
		"createClone":         miscfeatures.MiscCreateCloneTransformer.Transform,
		"createAssertClone":   miscfeatures.MiscCreateAssertCloneTransformer.Transform,
		"createIsClone":       miscfeatures.MiscCreateIsCloneTransformer.Transform,
		"createValidateClone": miscfeatures.MiscCreateValidateCloneTransformer.Transform,
	},
	"protobuf": {
		"message":              protobuffeatures.ProtobufMessageTransformer.Transform,
		"encode":               protobuffeatures.ProtobufEncodeTransformer.Transform,
		"assertEncode":         protobuffeatures.ProtobufAssertEncodeTransformer.Transform,
		"isEncode":             protobuffeatures.ProtobufIsEncodeTransformer.Transform,
		"validateEncode":       protobuffeatures.ProtobufValidateEncodeTransformer.Transform,
		"decode":               protobuffeatures.ProtobufDecodeTransformer.Transform,
		"assertDecode":         protobuffeatures.ProtobufAssertDecodeTransformer.Transform,
		"isDecode":             protobuffeatures.ProtobufIsDecodeTransformer.Transform,
		"validateDecode":       protobuffeatures.ProtobufValidateDecodeTransformer.Transform,
		"createEncode":         protobuffeatures.ProtobufCreateEncodeTransformer.Transform,
		"createAssertEncode":   protobuffeatures.ProtobufCreateAssertEncodeTransformer.Transform,
		"createIsEncode":       protobuffeatures.ProtobufCreateIsEncodeTransformer.Transform,
		"createValidateEncode": protobuffeatures.ProtobufCreateValidateEncodeTransformer.Transform,
		"createDecode":         protobuffeatures.ProtobufCreateDecodeTransformer.Transform,
		"createAssertDecode":   protobuffeatures.ProtobufCreateAssertDecodeTransformer.Transform,
		"createIsDecode":       protobuffeatures.ProtobufCreateIsDecodeTransformer.Transform,
		"createValidateDecode": protobuffeatures.ProtobufCreateValidateDecodeTransformer.Transform,
	},
	"notations": {
		"camel": notationfeatures.NotationGeneralTransformer.Transform("camel"),
	},
	"reflect": {
		"name":     reflectfeatures.ReflectNameTransformer.Transform,
		"metadata": reflectfeatures.ReflectMetadataTransformer.Transform,
		"schema":   reflectfeatures.ReflectSchemaTransformer.Transform,
		"schemas":  reflectfeatures.ReflectSchemasTransformer.Transform,
	},
}

func scanTypiaCalls(text string, roots []string) []internal.ITypiaCallExpression {
	var out []internal.ITypiaCallExpression
	for _, root := range roots {
		search := root + "."
		for offset := 0; ; {
			idx := strings.Index(text[offset:], search)
			if idx < 0 {
				break
			}
			start := offset + idx
			if start > 0 && isIdentifierPart(rune(text[start-1])) {
				offset = start + 1
				continue
			}
			if isInsideTriviaOrLiteral(text, start) {
				offset = start + len(search)
				continue
			}
			call, ok := parseTypiaCall(text, start, root)
			if ok {
				if len(call.Namespaces) == 0 || filepath.ToSlash(strings.Join(append(append([]string{}, call.Namespaces...), call.Method), "/")) == "llm/parameters" {
					call.TypeText = resolveLocalTypeText(text, call.TypeText)
				}
				out = append(out, call)
			}
			offset = start + len(search)
		}
	}
	return out
}

func resolveLocalTypeText(text string, typeText string) string {
	name := strings.TrimSpace(typeText)
	if name == "" || strings.ContainsAny(name, "<>|&[]{} ,") {
		return typeText
	}
	if name == "L1" {
		return typeText
	}
	token := "interface " + name
	idx := strings.Index(text, token)
	if idx < 0 {
		return typeText
	}
	open := strings.Index(text[idx:], "{")
	if open < 0 {
		return typeText
	}
	start := idx + open
	close, ok := matchBrace(text, start)
	if !ok {
		return typeText
	}
	return strings.TrimSpace(text[start : close+1])
}

func matchBrace(text string, pos int) (int, bool) {
	depth := 1
	for i := pos + 1; i < len(text); i++ {
		switch text[i] {
		case '{':
			depth++
		case '}':
			depth--
			if depth == 0 {
				return i, true
			}
		case '"', '\'', '`':
			q := text[i]
			for i++; i < len(text) && text[i] != q; i++ {
				if text[i] == '\\' {
					i++
				}
			}
		}
	}
	return 0, false
}

func isInsideTriviaOrLiteral(text string, pos int) bool {
	inLineComment := false
	inBlockComment := false
	quote := byte(0)
	for i := 0; i < pos && i < len(text); i++ {
		ch := text[i]
		if inLineComment {
			if ch == '\n' || ch == '\r' {
				inLineComment = false
			}
			continue
		}
		if inBlockComment {
			if ch == '*' && i+1 < pos && text[i+1] == '/' {
				inBlockComment = false
				i++
			}
			continue
		}
		if quote != 0 {
			if ch == '\\' {
				i++
				continue
			}
			if ch == quote {
				quote = 0
			}
			continue
		}
		if ch == '/' && i+1 < pos {
			if text[i+1] == '/' {
				inLineComment = true
				i++
				continue
			}
			if text[i+1] == '*' {
				inBlockComment = true
				i++
				continue
			}
		}
		if ch == '\'' || ch == '"' || ch == '`' {
			quote = ch
		}
	}
	return inLineComment || inBlockComment || quote != 0
}

func parseTypiaCall(text string, start int, root string) (internal.ITypiaCallExpression, bool) {
	i := start + len(root) + 1
	segments := []string{}
	for {
		name, next := readIdentifier(text, i)
		if name == "" {
			return internal.ITypiaCallExpression{}, false
		}
		segments = append(segments, name)
		i = next
		if i < len(text) && text[i] == '.' {
			i++
			continue
		}
		break
	}
	if i >= len(text) || text[i] != '<' {
		j := skipSpace(text, i)
		if j >= len(text) || text[j] != '(' {
			return internal.ITypiaCallExpression{}, false
		}
		method := segments[len(segments)-1]
		namespaces := append([]string(nil), segments[:len(segments)-1]...)
		return internal.ITypiaCallExpression{
			Root:       root,
			Namespaces: namespaces,
			Method:     method,
			TypeText:   "any",
			Factory:    isFactory(method, namespaces),
		}, true
	}
	closeType, ok := matchAngle(text, i)
	if !ok {
		return internal.ITypiaCallExpression{}, false
	}
	j := skipSpace(text, closeType+1)
	if j >= len(text) || text[j] != '(' {
		return internal.ITypiaCallExpression{}, false
	}
	method := segments[len(segments)-1]
	namespaces := append([]string(nil), segments[:len(segments)-1]...)
	return internal.ITypiaCallExpression{
		Root:       root,
		Namespaces: namespaces,
		Method:     method,
		TypeText:   strings.TrimSpace(text[i+1 : closeType]),
		Factory:    isFactory(method, namespaces),
	}, true
}

func isFactory(method string, namespaces []string) bool {
	if strings.HasPrefix(method, "create") {
		return true
	}
	switch filepath.ToSlash(strings.Join(append(namespaces, method), "/")) {
	case "json/schema", "json/schemas", "llm/application", "llm/parameters", "llm/structuredOutput", "misc/literals", "protobuf/message", "reflect/name", "reflect/metadata", "reflect/schema", "reflect/schemas":
		return true
	default:
		return false
	}
}

func readIdentifier(text string, i int) (string, int) {
	if i >= len(text) || !isIdentifierStart(rune(text[i])) {
		return "", i
	}
	j := i + 1
	for j < len(text) && isIdentifierPart(rune(text[j])) {
		j++
	}
	return text[i:j], j
}

func skipSpace(text string, i int) int {
	for i < len(text) {
		switch text[i] {
		case ' ', '\n', '\r', '\t':
			i++
		default:
			return i
		}
	}
	return i
}

func matchAngle(text string, pos int) (int, bool) {
	depth := 1
	for i := pos + 1; i < len(text); i++ {
		switch text[i] {
		case '<':
			depth++
		case '>':
			depth--
			if depth == 0 {
				return i, true
			}
		case '"', '\'', '`':
			q := text[i]
			for i++; i < len(text) && text[i] != q; i++ {
				if text[i] == '\\' {
					i++
				}
			}
		}
	}
	return 0, false
}

func isIdentifierStart(r rune) bool {
	return r == '_' || r == '$' || ('A' <= r && r <= 'Z') || ('a' <= r && r <= 'z')
}

func isIdentifierPart(r rune) bool {
	return isIdentifierStart(r) || ('0' <= r && r <= '9')
}
